### ISSUE TITLE
Raspberry Pi 4?

### ISSUE THREAD

**Issue by rgl** [ipxe/pipxe#2]
Do you have plans for making this work with Raspberry Pi 4?

**Comment by ipxe-devel** [comment]
On Mon, Dec 09, 2019 at 10:46:34AM -0800, Rui Lopes wrote:
> Do you have plans for making this work with Raspberry Pi 4?
> 

Please elaborate that question.


Groeten
Geert Stappers

P.S.

In case "issue #2" has that question better described
as the "you have plans",  then say so. I'll visit "git hub".
(This is a reply throught the iPXE mailinglist.
 Yes, there is a whole world outside "git hub".)
-- 
Leven en laten leven

**Comment by rgl** [comment]
I mean, does the `pipxe` project has plans for having iPXE running in the raspberry pi 4 firmware?

Currently, it seems there's beta support for PXE booting at https://github.com/Hexxeh/rpi-firmware, but how about iPXE? Are the two projects related? Working together?

Sorry, I didn't notice/look-for there was even a IRC channel / mailing list! Will ask there.

**Comment by mcb30** [comment]
They are completely unrelated projects.

piPXE relies on UEFI.  The UEFI support for the Pi 4 is in the process of being merged to the upstream edk2-platforms repositories.  Once that is done, it should be possible to invoke piPXE and have it work with any USB network adapters.

AFAIK the onboard NIC is (embedded) USB on the Pi 3B+ and earlier, but may be PCIe on the Pi 4.  If this is the case, then a suitable NIC driver for iPXE will need to be written.

**Comment by andreiw** [comment]
Hi folks.

https://rpi4-uefi.dev/ tracks the Pi 4 UEFI. GENET driver is developed and currently being cleaned up and soon to be upstreamed - and the expectation is that it should work with iPXE.

Feel free to join us on the #rpi4-uefi-dev channel on the Arm Developer-Ecosystem Discord server (https://discordapp.com/invite/fqRhc8y)

**Comment by blitmap** [comment]
So this looks promising:

https://rpi4-uefi.dev/v1-13-release-for-pi-4/

**Comment by rgl** [comment]
Forgot to update this issue... but I have to report that upstream ipxe (1.20.1) works fine with rpi4-uefi (you need v1.15 for rpi4+8gb support).

FYI, I have a vagrant environment for building ipxe and uefi at https://github.com/rgl/raspberrypi-uefi-edk2-vagrant.

**Comment by IntelMiner** [comment]
Apologies for the necro-post. But this was one of the first results in Google 

I've been tearing my hair out for several hours now trying to get a scripted iPXE to boot on my Raspberry Pi 4. I've tried using both the iPXE upstream, as well as the version 1.20.1 build that @rgl's repository provides 

No matter what version I build, executing a `make bin-arm64-efi/ipxe.efi EMBED=myscript.ipxe` or even building without the script builds a binary that boots, but then does not show any ethernet devices :(

I've got the latest EFI 1.24 firmware running, and have also tried 1.16 and 1.17 due to at least one thread noting that it may have errors netbooting 

I've found a few pre-made iPXE binaries that do already work, but I want to embed a script in to mine for HTTP booting

Is `make bin-arm64-efi/ipxe.efi` still the correct build command for the Raspberry Pi 4? Or am I missing some step somewhere

**Comment by IntelMiner** [comment]
Additional: While I was writing the above comment. My Pi 4 booted, gave a "net0 no such network device" error. Looped through the "Start PXE over IPv4" and then IPv6 

And then actually found net0 and attempted to boot!

Could this perhaps be a bug in either iPXE or the UEFI firmware I'm using?

**Comment by PeterSurda** [comment]
@IntelMiner I'm in the same situation as you. I think this project doesn't support Raspberry Pi 4. However, https://github.com/netbootxyz/netboot.xyz does. I'm now trying to find out how to build it with your own menu.

**Comment by PeterSurda** [comment]
So it looks like netboot.xyz does internally use pipxe, and researching did help me find out how to build pipxe for RPi4. Here is a condensed version:

1. manually update all submodules, in `edk2`, `edk2-platforms`, `edk2-non-osi` and do `git pull origin master`
2. edit the Makefile, replace `RPi3` with `RPi4`
3. also in the Makefile, disable git submodule update as it will then revert the update (just comment the line `git submodule update --init --recursive`)
4. also in the Makefile, change `IPXE_TGT` to `bin-arm64-efi/snp.efi`
4. run `make`

The resulting EFI file seems to work OK on my RPi4.

**Comment by IntelMiner** [comment]
I will give this a try in a few days and report back, thank you @PeterSurda!

**Comment by IntelMiner** [comment]
@PeterSurda Update!

It took some finesseing to get it to work on my build box (Gentoo needed the `iasl` package for those relevant files)

I ran into an ipxe build error with some junk relating to "infiniband" (possibly an issue with using GCC 10.2?) but commenting that out fixed it

Embedding my boot script seemed to work as well. It pulls down a kernel + initrd and (attempts) to boot them

However it now hangs at

`EFI stub: Booting Linux Kernel...`
`EFI stub: Generating empty DTB`
`EFI stub: generating boot services and installing virtual address map...`

But I know that's (probably) outside the scope of iPXE and likely a kernel configuration issue on my side

Thank you very much for all your help however :) Hopefully this thread helps others in future too

**Comment by stappersg** [comment]
> I ran into an ipxe build error with some junk relating to "infiniband" (possibly an issue with using GCC 10.2?) but commenting that out fixed it

Please elaborate (in another issue as this issue).

> Hopefully this thread helps others in future too

Yes, we aim for a better world.

**Comment by PeterSurda** [comment]
@IntelMiner Check here for my other posts/comments for more info:

- how to use MAAS images on Pi: https://discourse.maas.io/t/raspbery-pi-4-maas-images/4376/5?u=petersurda
- auto-boot PXEv4 from the UEFI firmware: https://github.com/tianocore/edk2-platforms/commit/f88bff4d000c69b1135969dbeafccaeb782dcdf3#r48792258

**Comment by OGKevin** [comment]
Im on macos and cant seem to build this for rpi4. I tried using a vargant box based on ubuntu but no luck there neither. Any pointers on how i could get this build for rpi4?

**Comment by IntelMiner** [comment]
Hi team,

I finally had some time to sit down and poke at this again.

I went ahead and wiped my old git copy and resynced from scratch just to make sure there weren't any lingering issues with testing. 

I upgraded to GCC 11.1.0 in the meantime, however I still seem to be getting the infiniband errors I mentioned previously @stappersg as well as another error, though the build continues after this first error in the "encode" library

Running a standard `make -j16` shows this error, but continues

```
enc/encode.c:1435:20: error: argument 5 of type 'const uint8_t *' {aka 'const unsigned char *'} declared as a pointer [-Werror=vla-parameter]
 1435 |     const uint8_t* input_buffer, size_t* encoded_size,
      |     ~~~~~~~~~~~~~~~^~~~~~~~~~~~
In file included from enc/encode.c:9:
./include/brotli/encode.h:293:19: note: previously declared as a variable length array 'const uint8_t[*encoded_size]' {aka 'const unsigned char[*encoded_size]'}
  293 |     const uint8_t input_buffer[BROTLI_ARRAY_PARAM(input_size)],
      |     ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
enc/encode.c:1436:14: error: argument 7 of type 'uint8_t *' {aka 'unsigned char *'} declared as a pointer [-Werror=vla-parameter]
 1436 |     uint8_t* encoded_buffer) {
      |     ~~~~~~~~~^~~~~~~~~~~~~~
In file included from enc/encode.c:9:
./include/brotli/encode.h:295:13: note: previously declared as a variable length array 'uint8_t[input_size]' {aka 'unsigned char[input_size]'}
  295 |     uint8_t encoded_buffer[BROTLI_ARRAY_PARAM(*encoded_size)]);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

After that it will switch to the bin-arm64-efi ipxe build, before eventually erroring out with

```
drivers/infiniband/flexboot_nodnic.c: In function 'flexboot_nodnic_create_qp':
drivers/infiniband/flexboot_nodnic.c:368:53: error: implicit conversion from 'enum ib_queue_pair_type' to 'nodnic_queue_pair_type' [-Werror=enum-conversion]
  368 |  status = nodnic_port_create_qp(&port->port_priv, qp->type,
      |                                                   ~~^~~~~~
drivers/infiniband/flexboot_nodnic.c: In function 'flexboot_nodnic_destroy_qp':
drivers/infiniband/flexboot_nodnic.c:409:45: error: implicit conversion from 'enum ib_queue_pair_type' to 'nodnic_queue_pair_type' [-Werror=enum-conversion]
  409 |  nodnic_port_destroy_qp(&port->port_priv, qp->type,
      |                                           ~~^~~~~~
```
